api/src/main/java/org/openmrs/module/queue/api/impl/QueueEntryServiceImpl.java

```diff
@@ -83,6 +84,19 @@ public QueueEntry saveQueueEntry(QueueEntry queueEntry) {
 				throw new IllegalArgumentException("Patient mismatch - visit.patient does not match patient");
 			}
 		}
+		QueueEntrySearchCriteria searchCriteria = new QueueEntrySearchCriteria();
+		searchCriteria.setPatient(queueEntry.getPatient());
+		searchCriteria.setQueues(Collections.singletonList(queueEntry.getQueue()));
+		List<QueueEntry> queueEntries = getQueueEntries(searchCriteria);
+		for (QueueEntry entry : queueEntries) {
+			if (!((queueEntry.getStartedAt().before(entry.getStartedAt())
+			        && (queueEntry.getEndedAt() != null && queueEntry.getEndedAt().before(entry.getStartedAt())))
+			        || ((entry.getEndedAt() != null) && (queueEntry.getStartedAt().after(entry.getEndedAt()))))) {
+				//if the new queueEntry endDate is NOT before the entry startDate OR the new queueEntry startDate is NOT after the entry endDate,
+				//then the new queueEntry overlaps with the existing entry
+				throw new DuplicateQueueEntryException("queue.entry.duplicate.patient");
```

Comment:
Will need to add the messageSourceService to this, wire it up via the application context xml, and then use it here to do the necessary translation.